### PR TITLE
Feat(GraphQL): This PR allows @id field in interface to be unique across all implementing types

### PR DESCRIPTION
**Pull request by harshil-goel**
Currently, @id fields in the interface are unique only in one implementing type. But there are several use cases that require @id field to be unique across all the implementation types. Also currently. get a query on the interface can result in unexpected errors as we can have multiple implementing types have the same value for that @id field.

Now we are allowing the @id field in the interface to be unique across all the implementing types. In order to do this, we have added a new argument interface of boolean type in the @id field.

Whenever a @id field in interface type has an interface argument set then its value will be unique across all the implementing types. Users will get errors if they try to add a node with such a field and there is already a node with the same value of that field even in some other implementing types. This is true for other scenarios like adding nested value or while using upserts.

If the interface argument is not present or its value is false then that field will be unique only for one implementing type. But such fields won't be allowed in argument to get query on interface in the future, see this PR also #7602

Example Schema,

 interface LibraryItem {
    refID: String! @id                   //  This field is unique only for one implementing type
    itemID: String! @id(interface:true)    // This field will be unique over all the implementing types inheriting this interface
}

type Book implements LibraryItem {
    title: String
    author: String
}
Related discuss Post: https://discuss.dgraph.io/t/uniqueness-for-id-fields-on-interface/13294

Cherry picked from: #7710

### CODE REVIEW

**Review by mangalaman93 — CHANGES_REQUESTED** [review]

File: graphql/e2e/auth/debug_off/debugoff_test.go [hunk 1]
```diff
@@ -122,6 +122,68 @@ func TestAddMutationWithXid(t *testing.T) {
...
+	gqlResponse = addSportsMemberParams.ExecuteAsPost(t, common.GraphqlURL)
+	common.RequireNoGQLErrors(t, gqlResponse)
+	var resultSportsMember struct {
+		AddSportsMember struct {
+			NumUids int
+		}
+	}
+	err = json.Unmarshal(gqlResponse.Data, &resultSportsMember)
+	require.NoError(t, err)
+	require.Equal(t, 0, resultSportsMember.AddSportsMember.NumUids)
```
[the remark this answers is not in the document]
Comment (harshil-goel) [reply graphql/e2e/auth/debug_off/debugoff_test.go:180]: It's an error in debug mode. Due to security reasons, we don't show an error in non debug mode.

File: graphql/e2e/directives/schema.graphql [hunk 2]
```diff
@@ -377,3 +377,35 @@ type Employer {
...
+
+type CricketTeam implements Team {
+    numOfBatsmen: Int
+    numOfBowlers: Int
+}
+
+type LibraryManager  {
+    name: String! @id
+    manages: [LibraryMember]
+}
```
Comment (mangalaman93) [inline graphql/e2e/directives/schema.graphql:411]: add a newline here

File: graphql/e2e/normal/schema.graphql [hunk 3]
```diff
@@ -380,3 +380,36 @@ type Dataset {
...
+
+type CricketTeam implements Team {
+        numOfBatsmen: Int
+        numOfBowlers: Int
+}
+
+type LibraryManager  {
+        name: String! @id
+        manages: [LibraryMember]
+}
```
Comment (mangalaman93) [inline graphql/e2e/normal/schema.graphql:415]: add a newline here

File: graphql/resolve/add_mutation_test.yaml [hunk 4]
```diff
@@ -1127,6 +1127,537 @@
...
+            "readHours": "4d2hr"
+        }
+    }
+  dgquery: |-
+    query {
+      LibraryMember_1(func: eq(Member.name, "Alice")) {
+        uid
+        dgraph.type
+      }
+      LibraryMember_2(func: eq(Member.refID, "101")) {
```
[the remark this answers is not in the document]
Comment (mangalaman93) [reply graphql/resolve/add_mutation_test.yaml:1160]: We should file a JIRA for this

File: graphql/resolve/mutation_rewriter.go [hunk 5]
```diff
@@ -131,9 +131,19 @@ func NewVariableGenerator() *VariableGenerator {
 func (v *VariableGenerator) Next(typ schema.Type, xidName, xidVal string, auth bool) string {
 	// return previously allocated variable name for repeating xidVal
 	var key string
+	flagAndXidName := xidName
+
+	// We pass the xidName as "Int.xidName" to generate variable for existence query
```
[the remark this answers is not in the document]
Comment (mangalaman93) [reply graphql/resolve/mutation_rewriter.go:136]: pending

File: graphql/resolve/mutation_rewriter.go [hunk 6]
```diff
@@ -1715,6 +1760,20 @@ func rewriteObject(
 	return frag, upsertVar, retErrors
 }
 
+func xidErrorForInterfaceType(typ schema.Type, xidString string, xid schema.FieldDefinition,
+	interfaceName string) error {
+	// TODO(GRAPHQL): currently we are checking typ of the mutated field for auth rules,
```
Comment (mangalaman93) [inline graphql/resolve/mutation_rewriter.go:1765]: q?

File: graphql/resolve/mutation_rewriter.go [hunk 7]
```diff
@@ -1406,46 +1418,79 @@ func rewriteObject(
...
+				_, interfaceUidExist := idExistence[interfaceVar]
+				typUid, typUidExist := idExistence[variable]
+
+				if interfaceUidExist || typUidExist {
 					// node with XID exists. This is a reference.
-					// We return an error if this is at toplevel. Else, we return the ID reference
+					// We return an error if this is at toplevel. Else, we return the ID reference if
+					// found node is of same type as xid field type. Because that node can be of some other
+					// type in case xidField is inherited from interface.
 					if atTopLevel {
```
Comment (mangalaman93) [inline graphql/resolve/mutation_rewriter.go:1444]: q?

File: graphql/resolve/mutation_rewriter.go [hunk 8]
```diff
@@ -2376,3 +2453,15 @@ func extractVal(xidVal interface{}, xidName, typeName string) (string, error) {
 			"allowed as Xid", xidName, typeName)
 	}
 }
+
+// This function will return interface type and variable for existence query on interface,
+// if given xid is inherited from interface, otherwise it will return nil and empty string
+func interfaceVariable(typ schema.Type, varGen *VariableGenerator, xidName string,
+	xidString string) (schema.Type, string) {
```
Comment (mangalaman93) [inline graphql/resolve/mutation_rewriter.go:2460]: nit: add a new line after multi line signature

File: graphql/resolve/mutation_rewriter.go [hunk 9]
```diff
@@ -1715,6 +1760,20 @@ func rewriteObject(
 	return frag, upsertVar, retErrors
 }
 
+func xidErrorForInterfaceType(typ schema.Type, xidString string, xid schema.FieldDefinition,
+	interfaceName string) error {
```
Comment (mangalaman93) [inline graphql/resolve/mutation_rewriter.go:1764]: nit: add a newline after multiline signature

File: graphql/schema/gqlschema.go [hunk 10]
```diff
@@ -1968,18 +1969,14 @@ func addGetQuery(schema *ast.Schema, defn *ast.Definition, providesTypeMap map[s
 			},
 		})
 	}
+
 	if hasXIDField {
-		if defn.Kind == "INTERFACE" {
```
Comment (mangalaman93) [inline graphql/schema/gqlschema.go:1972]: q?

**Review by mangalaman93 — CHANGES_REQUESTED** [review]

**Review by mangalaman93 — APPROVED** [review]
Please don't forget to file JIRA ticket for the pending work. LGTM

**Comment by all-seeing-code** [comment]
> . Also currently. get a query on the interface can result in unexpected errors as we can have multiple implementing types have the same value for that @id field.

Unclear

**Review by all-seeing-code — APPROVED** [review]

**Pull request merged**